swift/proto/swift_proto_library.bzl

```
@@ -0,0 +1,324 @@
+            include_dev_srch_paths = include_dev_srch_paths,
+            label = ctx.label,
+            linking_contexts = [
+                dep[CcInfo].linking_context
+                for dep in deps
+                if CcInfo in dep
+            ],
+            module_context = module_context,
+            swift_toolchain = swift_toolchain,
+        )
+    )
+
+    # Create the providers:
+    providers = [
+        DefaultInfo(
+            files = depset(compact([
+                module_context.swift.swiftdoc,
+                module_context.swift.swiftinterface,
+                module_context.swift.swiftmodule,
+                module_context.swift.swiftsourceinfo,
+                linking_output.library_to_link.static_library,
+                linking_output.library_to_link.pic_static_library,
+            ])),
+            runfiles = ctx.runfiles(
+                collect_data = True,
+                collect_default = True,
+                files = ctx.files.data,
+            ),
+        ),
+        OutputGroupInfo(**output_groups_from_other_compilation_outputs(
+            other_compilation_outputs = other_compilation_outputs,
+        )),
+        CcInfo(
+            compilation_context = module_context.clang.compilation_context,
+            linking_context = linking_context,
+        ),
+        swift_common.create_swift_info(
+            modules = [module_context],
+            swift_infos = get_providers(deps, SwiftInfo),
+        ),
+        SwiftProtoInfo(
+            module_name = module_name,
+            generated_swift_srcs = generated_swift_srcs,
+        ),
+    ]
+
+    # Propagate an `apple_common.Objc` provider with linking info about the
+    # library so that linking with Apple Starlark APIs/rules works correctly.
+    # TODO(b/171413861): This can be removed when the Obj-C rules are migrated
+    # to use `CcLinkingContext`.
+    providers.append(new_objc_provider(
+        additional_objc_infos = (
+            swift_toolchain.implicit_deps_providers.objc_infos
+        ),
+        deps = deps,
+        feature_configuration = feature_configuration,
+        is_test = ctx.attr.testonly,
+        module_context = module_context,
+        libraries_to_link = [linking_output.library_to_link],
+        swift_toolchain = swift_toolchain,
+    ))
+
+    return providers
+
+swift_proto_library = rule(
+    attrs = dicts.add(
+        swift_common.library_rule_attrs(
+            additional_deps_aspects = [
+                _swift_proto_library_aspect,
+                swift_clang_module_aspect,
+            ],
+            requires_srcs = False,
+        ),
+        {
+            "protos": attr.label_list(
+                doc = """\
+                Exactly one `proto_library` target (or a target producing `ProtoInfo`),
+                from which the Swift library should be generated.
+                """,
+                providers = [ProtoInfo],
+            ),
```

Thoughts on updating the name / type to make it more obvious its only one (or does it need to be a list)
```suggestion
            "dep": attr.label(
                doc = """\
                Exactly one `proto_library` target (or a target producing `ProtoInfo`),
                from which the Swift library should be generated.
                """,
                providers = [ProtoInfo],
            ),
```

---

No, actually I forgot do update this doc comment -- it actually works with an arbitrary number of proto targets!
I added the proto_files example demonstrating this in action. I'll update the doc comment.